### ISSUE TITLE
Configurable default mounts

### ISSUE THREAD

**Issue by townsend2010** [canonical/multipass#137]
It would be a great opportunity to allow something like a user setup of default mount.
So if a developer wants to test something from his `/tmp/testthis` it would be nice to not `multipass mount ...` every time.

Instead one could have some sort of `multipass mount --setdefault path path` and if set, mount those paths on every machine it spawns.

**Comment by andrei-toterman** [comment]
How about setting this through a key?
```
multipass set client.default-mount=<host-path>[:<instance-path>]
```
or something like that

**Comment by ricab** [comment]
Hey @andrei-toterman, yep I believe we were settling on a setting for this. There is a proposal (#756) that tucks it in with other launch defaults. I wonder if we should somehow articulate that with the existing default mount in primary :thinking: And would other specific defaults for primary make sense? Something to think about... WDYT?

**Comment by ricab** [comment]
Hey @andrei-toterman, have you thought of a solution you'd like for this? I would very much like to see proposals, or even just your thoughts on the matter.

Personally I would like to be able to disable the Home mount, as I find myself not using `primary` to save that extra time. I imagine other people would feel the same way (?)

**Comment by townsend2010** [comment]
Hey @ricab,

> Personally I would like to be able to disable the Home mount

I think you'd like to see https://github.com/canonical/multipass/issues/1861 fixed :slightly_smiling_face:

**Comment by ricab** [comment]
indeed

**Comment by andrei-toterman** [comment]
Hey, @ricab and @townsend2010!

Okay, so here are some options:
1. Have a simple `client.default-mount` key that applies to all instances and doesn't affect the default home mount of the primary instance i.e. it will mount both home and the specified path. This is the simplest option and the least confusing one. We could also make this key take a comma separated list of paths and mount all of them.
2. Have that `client.default-mount` key and make it override the primary home mount i.e. when the key is not set, primary mounts home, and when the key is set, all instances mount that path and primary doesn't mount home anymore. This is kind of confusing, in my opinion.
3. Like option 1 but in addition to that have a `client.primary-home-mount` boolean key that controls whether the primary instance mounts home or not. This is more explicit and configurable.
4. Like option 1 but have a `client.primary-mount` option that takes a path for the primary instance to mount and defaults to the current home mount. I think this is better option 3. 
5. The most comprehensive solution I could think of is a regex based one in which we can set up a list of pairs of regexes and associated mount paths for them, and have any instance whose name matches a regex to mount the associated paths. This would allow for basically limitless customization, but wouldn't fit in the context of simple key settings, only in a big config yaml file or something like that. Also this may simply be too much customization that won't really be used.

Now regarding the `launch --mount` option, I think this should be an override option. When it is not specified, the launched instance mounts according to the settings, but when it is specified, the instance mounts only the paths specified in it. Maybe we could also add a `--no-default-mounts` option when we want to launch a instance and not mount anything at all, but this would bloat the CLI. Or maybe handle `--mount false` separately and not mount anything if specified. But I think option 4 + a way to override and disable mounts for a specific instance at launch time through the `--mount` option would be configurable enough for most users.

**Comment by ricab** [comment]
Good insights @andrei-toterman, thanks. I'll think some more about this, but one thing I want to mention is that it should be clear from settings keys whether the mounts apply to existing instances or only to new launches. The proposal in #756 achieves that with `client.launch-defaults.mounts`. We should probably stick to something like that, with the view of expanding that sub-tree in the future (that's what I meant with forward-compatibility before). 

Also, I quite like the `mounts` branch taking a dictionary in that proposal, although we should probably stick to only mapping paths to begin with (i.e. no id maps yet). Comma-separated input would work reasonably well in the CLI then.

**Comment by ricab** [comment]
I agree that the current `multipass set` is not the best experience for multiple mounts at once and changing `set` that way would be useful. However, I see this as a generic problem whenever you want to configure a tree of things in one go. Handling non-leave branches of the configuration in `set` should be a separate task IMO.

BTW, we have had the goal of adding a `multipass config` command for a while. This would would give you access to the full configuration in your favorite editor, so that sort of overlaps with this.

**Comment by ricab** [comment]
Actually, now that I think of this, there is more to this feature than initially apparent (to me at least). A spec would have been well justified IMO. 

The structure of the config means that we need to break new ground in one of two ways:
  1. we either start taking non-leave branches as values in set (e.g. `mp set ...mounts='{"/local/path": "/remote/path"}'`)
  2. or we introduce dynamic keys (`mp set ...mounts./local/path=/remote/path`)

Ultimately, I suppose both approaches should be valid, but for this PR we should just pick one. The other can be done separately, in a later PR. 

Option 2 is probably the easiest to implement, but I suppose you already went with option 1 in your implementation. Hopefully it is well contained in the setting handler and we can add YAML parsing at a more generic level later on...

It would be interesting to investigate what other tools taking YAML configurations do about this (e.g. `git config set...`)

**Comment by ricab** [comment]
This would need `multipass config` and is just one of the defaults proposed in #756, so marking it duplicate of that one.
